### ISSUE TITLE
or_fun_call bad output with vec![]

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust-clippy#1667]
```
#![allow(unused_variables)]
fn main() {
    use std::collections::HashMap;
    let mut hash = HashMap::<u32, Vec<u32>>::new();
    let foo = hash.entry(1).or_insert(vec![]);
}
```

Gives:

```
warning: use of `or_insert` followed by a function call
 --> src\main.rs:5:15
  |
5 |     let foo = hash.entry(1).or_insert(vec![]);
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(or_fun_call)] on by default
help: try this
  |     let foo = hash.entry(1).or_insert_with(|| < [ _ ] > :: into_vec ( box [ $ ( $ x ) , * ] ));
  = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#or_fun_call
```

A nicer looking warning should show `.or_insert_with(|| vec![]);`

**Comment by clarcharr** [comment]
This potentially could recommend `or_insert_with(Vec::new)` as well.

**Comment by yacoob** [comment]
A variation on this:

```rust
warning: use of `or_insert` followed by a function call
  --> src/bin/04.rs:88:60
   |
88 |         let minutes = &mut minutes_frequency.entry(*guard).or_insert(vec![0; 60]);
   |                                                            ^^^^^^^^^^^^^^^^^^^^^^ help: try this: `or_insert_with(|| $ crate :: vec :: from_elem ( $ elem , $ n ))`
```

`|| vec![0; 60]` work just as well. :D I guess clippy expands the `vec!` macro, and then tries to advise on that.

**Comment by clarcharr** [comment]
in that case, you are actually allocating the `Vec` and should be using `or_insert_with`.

**Comment by yacoob** [comment]
Yup, that's what I've meant; updated the comment to be less vague.

**Comment by clarcharr** [comment]
Ah, gotcha!

**Comment by lzutao** [comment]
Current output:
```rust
warning: use of `or_insert` followed by a function call
 --> src/main.rs:5:29
  |
5 |     let foo = hash.entry(1).or_insert(vec![0; 60]);
  |                             ^^^^^^^^^^^^^^^^^^^^^^ help: try this: `or_insert_with(|| vec![0; 60])`
  |
  = note: `#[warn(clippy::or_fun_call)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
```

So this issue is fixed? @phansch
